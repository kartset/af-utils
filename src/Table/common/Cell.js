import React from "react";
import PropTypes from "prop-types";
import { getCellProps } from "../../utils/extraPropsGetters";

/*
    If all cells in a row would be completely empty - row can "collapse" short.
    To prevent it we can fill it with &nbsp;
*/
const DEFAULT_EMPTY_CELL_CONTENT = "\u00A0";

const Cell = ({ rowData, rowIndex, getCellData: getCellDataDefault, column, columnIndex, getCellExtraProps }) => {
    const { render, getEmptyCellData, dataKey, format, getCellData } = column;

    let cellData = rowData && ( getCellData || getCellDataDefault )( rowData, rowIndex, dataKey );
    
    if( cellData === undefined || cellData === "" ){
        cellData = getEmptyCellData ? getEmptyCellData( rowIndex, column ) : DEFAULT_EMPTY_CELL_CONTENT;
    }
    else{
        if( format ){
            cellData = format( cellData, rowData );
        }
        if( render ){
            cellData = render( cellData, rowData, column );
        }
    }

    return (
        <td {...getCellProps(rowData,columnIndex,columnIndex,getCellExtraProps)}>
            {cellData}
        </td>
    );
};

Cell.propTypes = {
    getCellData: PropTypes.func.isRequired,
    rowIndex: PropTypes.number.isRequired,
    columnIndex: PropTypes.number.isRequired,
    column: PropTypes.object.isRequired,
    rowData: PropTypes.object,
    getCellExtraProps: PropTypes.func
};

export default Cell;