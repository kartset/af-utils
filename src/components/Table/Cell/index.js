import PropTypes from "prop-types";

/*
    If all cells in a row would be completely empty - row can "collapse" short.
    To prevent it we can fill it with &nbsp;
*/
const DEFAULT_EMPTY_CELL_CONTENT = "\u00A0";

const Cell = ({ rowData, rowIndex, column, columnIndex, getCellExtraProps }) => {
    const { render, getEmptyCellData, dataKey, format, getCellData } = column;

    let cellData = rowData && getCellData( rowData, rowIndex, dataKey );
    
    if( cellData === undefined || cellData === "" ){
        cellData = getEmptyCellData ? getEmptyCellData( rowIndex, column ) : DEFAULT_EMPTY_CELL_CONTENT;
    }
    else{
        if( format ){
            cellData = format( cellData, rowData );
        }
        if( render ){
            cellData = render( cellData, rowData, column );
        }
    }

    return (
        <td {...( getCellExtraProps && getCellExtraProps(rowData,columnIndex,rowIndex))} aria-colindex={columnIndex + 1}>
            {cellData}
        </td>
    );
};

Cell.propTypes = {
    rowIndex: PropTypes.number.isRequired,
    columnIndex: PropTypes.number.isRequired,
    column: PropTypes.object.isRequired,
    rowData: PropTypes.object,
    getCellExtraProps: PropTypes.func
};

export default Cell;