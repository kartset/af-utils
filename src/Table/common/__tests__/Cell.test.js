import React from "react";
import { act } from "react-dom/test-utils";
import { render, unmountComponentAtNode } from "react-dom";
import pretty from "pretty";
import Cell from "../Cell";

let container = null;

beforeEach(() => {
	container = document.createElement("tr");
	document.body.appendChild(container);
});

afterEach(() => {
	unmountComponentAtNode(container);
	container.remove();
	container = null;
});

const getCellData = ( rowData, rowIndex, dataKey ) => rowData[ dataKey ];

describe("Table DefaultCellComponent renders ok", () => {
	test("Performs initial render with default props", () => {
		act(() => {
			render(
				<Cell
					rowData={{ key1: "data1", key2: "data2" }}
					getCellData={getCellData}
					rowIndex={1}
					column={{ dataKey: "key1" }}
					columnIndex={1}
				/>,
				container
			);
		});

		expect(pretty(container.innerHTML)).toMatchSnapshot();
    });
    
    test("renders &nbsp; when cellData is empty", () => {
        act(() => {
			render(
				<Cell
					rowData={{}}
					getCellData={getCellData}
					rowIndex={1}
					column={{ dataKey: "key1" }}
					columnIndex={1}
				/>,
				container
			);
		});

		expect(pretty(container.innerHTML)).toMatchSnapshot();
    });
});
