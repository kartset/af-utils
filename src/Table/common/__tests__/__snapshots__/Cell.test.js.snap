// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table DefaultCellComponent renders ok Performs initial render with default props 1`] = `"<td aria-colindex=\\"2\\">data1</td>"`;

exports[`Table DefaultCellComponent renders ok renders &nbsp; when cellData is empty 1`] = `"<td aria-colindex=\\"2\\">&nbsp;</td>"`;
